### ISSUE TITLE
Detected as malware by banking app

### ISSUE THREAD

**Issue by CurrentSubset68** [florisboard/florisboard#2658]
### Short description

Banking app refuses to allow access with floris board set as the current keyboard. It can be worked around by switching to the stock keyboard when using the app. The banking app names floris board specifically. I understand if this is unfixable, banking apps can be quite strict with trusted computing.
Also not willing to name my bank, sorry

### Steps to reproduce

1. enable floris board
2. launch banking app
3. denies access stating a security risk from the app

### FlorisBoard Version

0.4.1

### Install Source

F-Droid

### Device

Pixel 7a

### Android

14, calyxOS 5.12.0

**Comment by CurrentSubset68** [comment]
Doesn't happen with AOSP keyboard, and it only started happening after a very recent update to the banking app. It was working fine before. I'll try with some other keyboards as well

**Comment by CurrentSubset68** [comment]
I have safety net enabled through micro g so i should hopefully be good for a while.

**Comment by VGHS-lucaruby** [comment]
I have had encountered the same issue.

I spoke with HSBC and they have stated that this is a new policy change as they were experiencing fraud cases involving non default keyboards.
They have logged it as feedback regarding FlorisBoard being incorrectly detected but I strongly doubt anything will come of it.
I would suggest also contacting them and ask to give feedback on this policy change and crossing your fingers.

As for it being a integrity issue I am using a Samsung S24 Ultra non rooted nor custom ROM. They are purely targeting non standard keyboards.

Can only assume AOSP keyboard working is due to it having been whitelisted either due to popularity or a developer involved happening to use it.

**Comment by VGHS-lucaruby** [comment]
I have had encountered the same issue.

I spoke with the bank in question and they have stated that this is a new policy change as they were experiencing fraud cases involving non default keyboards.
They have logged it as feedback regarding FlorisBoard being incorrectly detected but I strongly doubt anything will come of it.
I would suggest also contacting them and ask to give feedback on this policy change and crossing your fingers.

As for it being a integrity issue I am using a Samsung S24 Ultra non rooted nor custom ROM. They are purely targeting non standard keyboards.

Can only assume AOSP keyboard working is due to it having been whitelisted either due to popularity or a developer involved happening to use it.

**Comment by CurrentSubset68** [comment]
I can kinda understand where they're coming from, because malware is rampant on the playstore, as well as rouge apps that act genuine for a short while, but open source keyboards can be reviewed so arguably these should be whitelisted.

**Comment by VGHS-lucaruby** [comment]
Yeah, I fully agree. That is pretty much the feedback I gave them.
I have little hope they will fix it but maybe if enough users complain they may increase the whitelist.
My expirence with them on other matters has been quite good so I can only hope.

**Comment by CurrentSubset68** [comment]
Response from my bank after the automated messages was "hopefully this would be a part of the next major version" lmao
Lawyer wording

**Comment by VGHS-lucaruby** [comment]
Ah thats odd.
I phoned them so maybe thats why I got a different response.

**Comment by VGHS-lucaruby** [comment]
> 
> I can't. They are taking advantage of the fact android allows making such restrictions, something that can't be done in say mobile browser or desktop OS to mess with people or further control. It does not help, and waste their time (and possibly business) in the meantime. Plus Florisboard has no network permission, stock AOSP keyboard in Pixel (GBoard) does, plus sends things to google, has trackers and is proprietary. And they want customers to use the less secure and less private GBoard? That's weird lol
> 

This was my exact argument, so odd to ask users to have worse security lol

> Like blocking rooted devices - Windows, macOS and linux desktop OS's are rooted out of the box, so will they just block their main website, for security reasons?
>

Funniest part is there solution outside of changing keyboard was to use the website haha

**Comment by VGHS-lucaruby** [comment]
> I am not sure how they would come to that conclusion, sounds more like an excuse from the bank to avoid revealing the real reason.

Yeah I would love to see what caused them to reach this conclusion.

> Only security issue I could see in a 3rd party keyboard is if it is malicious (e.g backdoored by dev to send keystrokes to remote server). If they have evidence of customers that had fraud commited against them and all those customers reported using same keyboard, i could see it but then the question would be, why not block that specific keyboard only?

This was my only thought as to why

> Florisboard is not only open source, but has no network permission and therefore can't send anything to the internet, so should be no reason to block it.

Yep :(

**Comment by VGHS-lucaruby** [comment]
> Probably not in a feasible way.

Yeah, why i have placed most of my effort into contacting them everyway I can and attempt to explain how counter productive the change is.

**Comment by patrickgold** [comment]
> I spoke with the bank in question and they have stated that this is a new policy change as they were experiencing fraud cases involving non default keyboards.

What would be interesting to know is if they operate on a whitelist or blacklist system. Because I'd imagine that they didn't specifically block FlorisBoard and HeliBoard, but just created a whitelist of some known keyboards they think are safe and called it a day.

To verify this theory though users using the banking apps that block 3rd party keyboards would need to perform this test on their device. It is pretty simple: clone the project locally, check out the v0.4.3 tag, then change the `applicationId` in `app/build.gradle.kts` to something random, lets say `org.test.keyboard` and build/install that via adb. If the banking app still blocks this they use a whitelist, if it works then they operate on a blacklist.

**Comment by patrickgold** [comment]
> That's what I originally thought, but it seems the banking app is blocking it when it is currently active, not whether is is installed or not.

Yes that's what I think too, what I want to know though if the currently active IME id is matched against a whitelist or blacklist.

> Which lead me to wonder if they are using an API that ask the OS if the default keyboard (or even a specific list of their approved ones) is currently active. [...] Do you know if such an API is available?

You can very easily get the list of all enabled keyboards or of the currently active keyboard, see here for the system API:

https://github.com/florisboard/florisboard/blob/a89af25eab67b60d5594fa8ce828cae87004c800/app/src/main/kotlin/dev/patrickgold/florisboard/lib/util/InputMethodUtils.kt#L44-L68

**Comment by VGHS-lucaruby** [comment]
My assumption is they are operating a whitelist but I will give building a version with a different id a try when I get a minute this weekend.

**Comment by CurrentSubset68** [comment]
If it is a whitelist, would it be possible to spoof it as the AOSP keyboard? If the block is by package name it might be possible

**Comment by VGHS-lucaruby** [comment]
I was also thinking about trying this but if that works this change will be even stupider haha

**Comment by CurrentSubset68** [comment]
I dunno, proprietary source code can be pretty wacky

**Comment by VGHS-lucaruby** [comment]
Finally had some time to test.
Unfortunately swapping out the `applicationId` did not bypass this block.
I can only assume it is a whitelist based off of signatures or some other google play check. :(
If anyone has any other ideas to test let me know.

**Comment by MattRC7** [comment]
My bank is a bit more reasonable and simply warns me each time I open the app. Annoying, sure, but not a big deal. Maybe you'd have more luck convincing them to change it to a warning?

![The text reads, Warning, you appear to be using a third-party keyboard which is risky because it may be collecting all the text you type, including personal data like passwords and credit card numbers. We recommens you use your device default keyboard instead.](https://github.com/user-attachments/assets/c4d8a819-4c63-4b1c-b791-0e640debc38c)

**Comment by lm41** [comment]
I'm removing the bug label as it doesn't seems to be a florisboard bug. Sadly there is nothing we can do :(
